README.md
# ngWebSocket

A small JavaScript library that decorates the WebSocket API to provide a WebSocket connection that will automatically reconnect if the connection is dropped.

It is API compatible, so when you have:

```javascript
var ws = new WebSocket('ws://....');
```

you can replace with:

```javascript
var ws = new $webSocket('ws://....');
```

Minified library with gzip compression is less than 600 bytes.

## Installing

You can install this package locally with `bower`.

> Please note that Angular WebSocket requires **Angular 1.3.x** or higher.

```shell
# To get the latest stable version, use bower from the command line.
bower install angular-rws

# To get the most recent, last committed-to-master version use:
bower install 'angular-rws#master'

# To save the bower settings for future use:
bower install angular-rws --save

# Later, you can use easily update with:
bower update
```

## Using the library

Now that you have installed the Angular libraries, simply include the scripts and stylesheet in your main HTML file, in the order shown in the example below. Note that npm  will install the files under `/bower_components/angular-material/`.

```html
<!DOCTYPE html>
<html>
<head>
</head>
    <body ng-app="YourApp">
    <div ng-controller="YourController">
    </div>

    <script src="/bower_components/angular/angular.js"></script>
    <script src="/bower_components/angular-material/angular-rws.js"></script>
    <script>

        // Include app dependency on ngWebSocket

        angular.module( 'YourApp', [ 'ngWebSocket' ] )
            .controller("YourController", YourController );

    </script>

</body>
</html>
```

## How reconnections occur

With the standard `WebSocket` API, the events you receive from the WebSocket instance are typically:

    onopen
    onmessage
    onmessage
    onmessage
    onclose // At this point the WebSocket instance is dead.

With a `$webSocket`, after an `onclose` event is called it will automatically attempt to reconnect. In addition, a connection is attempted repeatedly (with a small pause) until it succeeds. So the events you receive may look something more like:

    onopen
    onmessage
    onmessage
    onmessage
    onclose
    // Attempts to reconnect
    onopen
    onmessage
    onmessage
    onmessage
    onclose
    // Attempts to reconnect
    onopen
    onmessage
    onmessage
    onmessage
    onclose

This is all handled automatically for you by the library.

### Parameters

```javascript
var socket = new $webSocket(url, protocols, options);
```

#### `url`

- The URL you are connecting to.
- <https://html.spec.whatwg.org/multipage/comms.html#network>

#### `protocols`

- Optional string or array of protocols per the WebSocket spec.
- <https://tools.ietf.org/html/rfc6455>

#### `options`

- Options (see below)

### Options

Options can either be passed as the 3rd parameter upon instantiation or set directly on the object after instantiation:

```javascript
var socket = new $webSocket(url, null, {debug: true, reconnectInterval: 3000});
```

or

```javascript
var socket = new $webSocket(url);
socket.debug = true;
socket.timeout_interval = 5400;
```

#### `debug`

- Whether this instance should log debug messages or not. Debug messages are printed to `console.debug()`.
- Accepts `true` or `false`
- Default value: `false`

#### `automatic_open`

- Whether or not the websocket should attempt to connect immediately upon instantiation. The socket can be manually opened or closed at any time using ws.open() and ws.close().
- Accepts `true` or `false`
- Default value: `true`

#### `reconnect_interval`

- The number of milliseconds to delay before attempting to reconnect.
- Accepts `integer`
- Default: `1000`

#### `max_reconnect_interval`

- The maximum number of milliseconds to delay a reconnection attempt.
- Accepts `integer`
- Default: `30000`

#### `reconnect_decay`

- The rate of increase of the reconnect delay. Allows reconnect attempts to back off when problems persist.
- Accepts `integer` or `float`
- Default: `1.5`

#### `timeout_interval`

- The maximum time in milliseconds to wait for a connection to succeed before closing and retrying.
- Accepts `integer`
- Default: `2000`

#### `max_reconnect_attempts`

- The maximum number of reconnection attempts that will be made before giving up. If null, reconnection attempts will be continue to be made forever.
- Accepts `integer` or `null`.
- Default: `null`

#### `binary_type`

- The binary type is required by some applications.
- Accepts strings `'blob'` or `'arraybuffer'`.
- Default: `'blob'`

---

### Methods

#### `ws.open()`

- Open the Reconnecting Websocket

#### `ws.close(code, reason)`

- Closes the WebSocket connection or connection attempt, if any. If the connection is already CLOSED, this method does nothing.
- `code` is optional the closing code (default value 1000). [https://tools.ietf.org/html/rfc6455#section-7.4.1](https://tools.ietf.org/html/rfc6455#section-7.4.1)
- `reason` is the optional reason that the socket is being closed. [https://tools.ietf.org/html/rfc6455#section-7.1.6](https://tools.ietf.org/html/rfc6455#section-7.1.6)

#### `ws.refresh()`

- Refresh the connection if still open (close and then re-open it).

#### `ws.send(data)`

- Transmits data to the server over the WebSocket connection.
- Accepts @param data a text string, ArrayBuffer or Blob

Like this? Check out [angular-rws](https://github.com/ajsb85/angular-rws) for the simplest way to create WebSocket backends from any programming language.

CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="1.0.0"></a>
# 1.0.0 (2017-01-05)

angular-websocket.js
(function () {
  'use strict';

  angular.module("ngWebsocket", [])
    .factory('reconnectingWebSocket', ['$window', '$document', '$timeout', '$log', function ($window, $document, $timeout, $log) {

      function ReconnectingWebSocket(url, protocols, options) {

        // Default settings
        var settings = {

          /** Whether this instance should log debug messages. */
          debug: false,

          /** Whether or not the $window.WebSocket should attempt to connect immediately upon instantiation. */
          automatic_open: true,

          /** The number of milliseconds to delay before attempting to reconnect. */
          reconnect_interval: 1000,
          /** The maximum number of milliseconds to delay a reconnection attempt. */
          max_reconnect_interval: 30000,
          /** The rate of increase of the reconnect delay. Allows reconnect attempts to back off when problems persist. */
          reconnect_decay: 1.5,

          /** The maximum time in milliseconds to wait for a connection to succeed before closing and retrying. */
          timeout_interval: 2000,

          /** The maximum number of reconnection attempts to make. Unlimited if null. */
          max_reconnect_attempts: null
        };
        if (!options) { options = {}; }

        // Overwrite and define settings with options if they exist.
        for (var key in settings) {
          if (typeof options[key] !== 'undefined') {
            this[key] = options[key];
          } else {
            this[key] = settings[key];
          }
        }

        // These should be treated as read-only properties

        /** The URL as resolved by the constructor. This is always an absolute URL. Read only. */
        this.url = url;

        /** The number of attempted reconnects since starting, or the last successful connection. Read only. */
        this.reconnectAttempts = 0;

        /**
         * The current state of the connection.
         * Can be one of: $window.WebSocket.CONNECTING, $window.WebSocket.OPEN, $window.WebSocket.CLOSING, $window.WebSocket.CLOSED
         * Read only.
         */
        this.readyState = $window.WebSocket.CONNECTING;

        /**
         * A string indicating the name of the sub-protocol the server selected; this will be one of
         * the strings specified in the protocols parameter when creating the $window.WebSocket object.
         * Read only.
         */
        this.protocol = null;

        // Private state variables

        var self = this;
        var ws;
        var timeout;
        var timed_out = false;
        var forced_close = false;
        var eventTarget = document.createElement('div');

        // Wire up "on*" properties as event handlers

        eventTarget.addEventListener('open', function (event) { self.onopen(event); });
        eventTarget.addEventListener('close', function (event) { self.onclose(event); });
        eventTarget.addEventListener('connecting', function (event) { self.onconnecting(event); });
        eventTarget.addEventListener('message', function (event) { self.onmessage(event); });
        eventTarget.addEventListener('error', function (event) { self.onerror(event); });

        // Expose the API required by EventTarget

        this.addEventListener = eventTarget.addEventListener.bind(eventTarget);
        this.removeEventListener = eventTarget.removeEventListener.bind(eventTarget);
        this.dispatchEvent = eventTarget.dispatchEvent.bind(eventTarget);

        /**
         * This function generates an event that is compatible with standard
         * compliant browsers and IE9 - IE11
         *
         * This will prevent the error:
         * Object doesn't support this action
         *
         * http://stackoverflow.com/questions/19345392/why-arent-my-parameters-getting-passed-through-to-a-dispatched-event/19345563#19345563
         * @param s String The name that the event should use
         * @param args Object an optional object that the event will use
         */
        function generateEvent(s, args) {
          var evt = document.createEvent("CustomEvent");
          evt.initCustomEvent(s, false, false, args);
          return evt;
        }

        this.open = function (reconnect_attempt) {
          ws = new $window.WebSocket(self.url, protocols || []);

          if (reconnect_attempt) {
            if (this.max_reconnect_attempts && this.reconnectAttempts > this.max_reconnect_attempts) {
              return;
            }
          } else {
            eventTarget.dispatchEvent(generateEvent('connecting'));
            this.reconnectAttempts = 0;
          }

          if (self.debug || ReconnectingWebSocket.debugAll) {
            $log.debug('ReconnectingWebSocket', 'attempt-connect', self.url);
          }

          var localWs = ws;
          timeout = $timeout(function () {
            if (self.debug || ReconnectingWebSocket.debugAll) {
              $log.debug('ReconnectingWebSocket', 'connection-timeout', self.url);
            }
            timed_out = true;
            localWs.close();
            timed_out = false;
          }, self.timeout_interval);

          ws.onopen = function (event) {
            $timeout.cancel(timeout);
            if (self.debug || ReconnectingWebSocket.debugAll) {
              $log.debug('ReconnectingWebSocket', 'onopen', self.url);
            }
            self.protocol = ws.protocol;
            self.readyState = $window.WebSocket.OPEN;
            self.reconnectAttempts = 0;
            var e = generateEvent('open');
            e.isReconnect = reconnect_attempt;
            reconnect_attempt = false;
            eventTarget.dispatchEvent(e);
          };

          ws.onclose = function (event) {
            $timeout.cancel(timeout);
            ws = null;
            if (forced_close) {
              self.readyState = $window.WebSocket.CLOSED;
              eventTarget.dispatchEvent(generateEvent('close'));
            } else {
              self.readyState = $window.WebSocket.CONNECTING;
              var e = generateEvent('connecting');
              e.code = event.code;
              e.reason = event.reason;
              e.wasClean = event.wasClean;
              eventTarget.dispatchEvent(e);
              if (!reconnect_attempt && !timed_out) {
                if (self.debug || ReconnectingWebSocket.debugAll) {
                  $log.debug('ReconnectingWebSocket', 'onclose', self.url);
                }
                eventTarget.dispatchEvent(generateEvent('close'));
              }

              timeout = self.reconnect_interval * Math.pow(self.reconnect_decay, self.reconnectAttempts);
              $timeout(function () {
                self.reconnectAttempts++;
                self.open(true);
              }, timeout > self.max_reconnect_interval ? self.max_reconnect_interval : timeout);
            }
          };
          ws.onmessage = function (event) {
            if (self.debug || ReconnectingWebSocket.debugAll) {
              $log.debug('ReconnectingWebSocket', 'onmessage', self.url, event.data);
            }
            var e = generateEvent('message');
            e.data = event.data;
            eventTarget.dispatchEvent(e);
          };
          ws.onerror = function (event) {
            if (self.debug || ReconnectingWebSocket.debugAll) {
              $log.debug('ReconnectingWebSocket', 'onerror', self.url, event);
            }
            eventTarget.dispatchEvent(generateEvent('error'));
          };
        };

        // Whether or not to create a $window.WebSocket upon instantiation
        if (this.automatic_open === true) {
          this.open(false);
        }

        /**
         * Transmits data to the server over the $window.WebSocket connection.
         *
         * @param data a text string, ArrayBuffer or Blob to send to the server.
         */
        this.send = function (data) {
          if (ws) {
            if (self.debug || ReconnectingWebSocket.debugAll) {
              $log.debug('ReconnectingWebSocket', 'send', self.url, data);
            }
            return ws.send(data);
          } else {
            throw 'INVALID_STATE_ERR : Pausing to reconnect $window.WebSocket';
          }
        };

        /**
         * Closes the $window.WebSocket connection or connection attempt, if any.
         * If the connection is already CLOSED, this method does nothing.
         */
        this.close = function (code, reason) {
          // Default CLOSE_NORMAL code
          if (typeof code == 'undefined') {
            code = 1000;
          }
          forced_close = true;
          if (ws) {
            ws.close(code, reason);
          }
        };

        /**
         * Additional public API method to refresh the connection if still open (close, re-open).
         * For example, if the app suspects bad data / missed heart beats, it can try to refresh.
         */
        this.refresh = function () {
          if (ws) {
            ws.close();
          }
        };
      }

      /**
       * An event listener to be called when the $window.WebSocket connection's readyState changes to OPEN;
       * this indicates that the connection is ready to send and receive data.
       */
      ReconnectingWebSocket.prototype.onopen = function (event) { };
      /** An event listener to be called when the $window.WebSocket connection's readyState changes to CLOSED. */
      ReconnectingWebSocket.prototype.onclose = function (event) { };
      /** An event listener to be called when a connection begins being attempted. */
      ReconnectingWebSocket.prototype.onconnecting = function (event) { };
      /** An event listener to be called when a message is received from the server. */
      ReconnectingWebSocket.prototype.onmessage = function (event) { };
      /** An event listener to be called when an error occurs. */
      ReconnectingWebSocket.prototype.onerror = function (event) { };

      /**
       * Whether all instances of ReconnectingWebSocket should log debug messages.
       * Setting this to true is the equivalent of setting all instances of ReconnectingWebSocket.debug to true.
       */
      ReconnectingWebSocket.debugAll = false;

      ReconnectingWebSocket.CONNECTING = $window.WebSocket.CONNECTING;
      ReconnectingWebSocket.OPEN = $window.WebSocket.OPEN;
      ReconnectingWebSocket.CLOSING = $window.WebSocket.CLOSING;
      ReconnectingWebSocket.CLOSED = $window.WebSocket.CLOSED;

      return ReconnectingWebSocket;

    }]);
})();

angular-rws.js
(function (angular) {
  'use strict';

  angular.module("ngWebSocket", [])
    .factory('$webSocket', ['$window', '$document', '$timeout', '$log', function ($window, $document, $timeout, $log) {

      function WebSocket(url, protocols, options) {

        // Default settings
        var settings = {

          /** Whether this instance should log debug messages. */
          debug: false,

          /** Whether or not the $window.WebSocket should attempt to connect immediately upon instantiation. */
          automatic_open: true,

          /** The number of milliseconds to delay before attempting to reconnect. */
          reconnect_interval: 1000,
          /** The maximum number of milliseconds to delay a reconnection attempt. */
          max_reconnect_interval: 30000,
          /** The rate of increase of the reconnect delay. Allows reconnect attempts to back off when problems persist. */
          reconnect_decay: 1.5,

          /** The maximum time in milliseconds to wait for a connection to succeed before closing and retrying. */
          timeout_interval: 2000,

          /** The maximum number of reconnection attempts to make. Unlimited if null. */
          max_reconnect_attempts: null
        };
        if (!options) { options = {}; }

        // Overwrite and define settings with options if they exist.
        for (var key in settings) {
          if (typeof options[key] !== 'undefined') {
            this[key] = options[key];
          } else {
            this[key] = settings[key];
          }
        }

        // These should be treated as read-only properties

        /** The URL as resolved by the constructor. This is always an absolute URL. Read only. */
        this.url = url;

        /** The number of attempted reconnects since starting, or the last successful connection. Read only. */
        this.reconnect_attempts = 0;

        /**
         * The current state of the connection.
         * Can be one of: $window.WebSocket.CONNECTING, $window.WebSocket.OPEN, $window.WebSocket.CLOSING, $window.WebSocket.CLOSED
         * Read only.
         */
        this.readyState = $window.WebSocket.CONNECTING;
        
        /**
         * A string indicating the name of the sub-protocol the server selected; this will be one of
         * the strings specified in the protocols parameter when creating the $window.WebSocket object.
         * Read only.
         */
        this.protocol = null;

        // Private state variables

        var self = this;
        var ws;
        var timeout;
        var timed_out = false;
        var forced_close = false;
        //var eventTarget = angular.element('<div></div>');
        var eventTarget = document.createElement('div'); 

        // Wire up "on*" properties as event handlers

        eventTarget.addEventListener('open', function (event) { self.onopen(event); });
        eventTarget.addEventListener('close', function (event) { self.onclose(event); });
        eventTarget.addEventListener('connecting', function (event) { self.onconnecting(event); });
        eventTarget.addEventListener('message', function (event) { self.onmessage(event); });
        eventTarget.addEventListener('error', function (event) { self.onerror(event); });

        // Expose the API required by EventTarget

        this.addEventListener = eventTarget.addEventListener.bind(eventTarget);
        this.removeEventListener = eventTarget.removeEventListener.bind(eventTarget);
        this.dispatchEvent = eventTarget.dispatchEvent.bind(eventTarget);

        /**
         * This function generates an event that is compatible with standard
         * compliant browsers and IE9 - IE11
         *
         * This will prevent the error:
         * Object doesn't support this action
         *
         * http://stackoverflow.com/questions/19345392/why-arent-my-parameters-getting-passed-through-to-a-dispatched-event/19345563#19345563
         * @param s String The name that the event should use
         * @param args Object an optional object that the event will use
         */
        function generateEvent(s, args) {
          var evt = document.createEvent("CustomEvent");
          evt.initCustomEvent(s, false, false, args);
          return evt;
        }

        this.open = function (reconnect_attempt) {
          ws = new $window.WebSocket(self.url, protocols || []);

          if (reconnect_attempt) {
            if (this.max_reconnect_attempts && this.reconnect_attempts > this.max_reconnect_attempts) {
              return;
            }
          } else {
            eventTarget.dispatchEvent(generateEvent('connecting'));
            this.reconnect_attempts = 0;
          }

          if (self.debug || WebSocket.debug_all) {
            $log.debug('WebSocket', 'attempt-connect', self.url);
          }

          var localWs = ws;
          timeout = $timeout(function () {
            if (self.debug || WebSocket.debug_all) {
              $log.debug('WebSocket', 'connection-timeout', self.url);
            }
            timed_out = true;
            localWs.close();
            timed_out = false;
          }, self.timeout_interval);

          ws.onopen = function (event) {
            $timeout.cancel(timeout);
            if (self.debug || WebSocket.debug_all) {
              $log.debug('WebSocket', 'onopen', self.url);
            }
            self.protocol = ws.protocol;
            self.readyState = $window.WebSocket.OPEN;
            self.reconnect_attempts = 0;
            var e = generateEvent('open');
            e.isReconnect = reconnect_attempt;
            reconnect_attempt = false;
            eventTarget.dispatchEvent(e);
          };

          ws.onclose = function (event) {
            $timeout.cancel(timeout);
            ws = null;
            if (forced_close) {
              self.readyState = $window.WebSocket.CLOSED;
              eventTarget.dispatchEvent(generateEvent('close'));
            } else {
              self.readyState = $window.WebSocket.CONNECTING;
              var e = generateEvent('connecting');
              e.code = event.code;
              e.reason = event.reason;
              e.wasClean = event.wasClean;
              eventTarget.dispatchEvent(e);
              if (!reconnect_attempt && !timed_out) {
                if (self.debug || WebSocket.debug_all) {
                  $log.debug('WebSocket', 'onclose', self.url);
                }
                eventTarget.dispatchEvent(generateEvent('close'));
              }

              timeout = self.reconnect_interval * Math.pow(self.reconnect_decay, self.reconnect_attempts);
              $timeout(function () {
                self.reconnect_attempts++;
                self.open(true);
              }, timeout > self.max_reconnect_interval ? self.max_reconnect_interval : timeout);
            }
          };
          ws.onmessage = function (event) {
            if (self.debug || WebSocket.debug_all) {
              $log.debug('WebSocket', 'onmessage', self.url, event.data);
            }
            var e = generateEvent('message');
            e.data = event.data;
            eventTarget.dispatchEvent(e);
          };
          ws.onerror = function (event) {
            if (self.debug || WebSocket.debug_all) {
              $log.debug('WebSocket', 'onerror', self.url, event);
            }
            eventTarget.dispatchEvent(generateEvent('error'));
          };
        };

        // Whether or not to create a $window.WebSocket upon instantiation
        if (this.automatic_open === true) {
          this.open(false);
        }

        /**
         * Transmits data to the server over the $window.WebSocket connection.
         *
         * @param data a text string, ArrayBuffer or Blob to send to the server.
         */
        this.send = function (data) {
          if (ws) {
            if (self.debug || WebSocket.debug_all) {
              $log.debug('WebSocket', 'send', self.url, data);
            }
            return ws.send(data);
          } else {
            throw 'INVALID_STATE_ERR : Pausing to reconnect $window.WebSocket';
          }
        };

        /**
         * Closes the $window.WebSocket connection or connection attempt, if any.
         * If the connection is already CLOSED, this method does nothing.
         */
        this.close = function (code, reason) {
          // Default CLOSE_NORMAL code
          if (typeof code == 'undefined') {
            code = 1000;
          }
          forced_close = true;
          if (ws) {
            ws.close(code, reason);
          }
        };

        /**
         * Additional public API method to refresh the connection if still open (close, re-open).
         * For example, if the app suspects bad data / missed heart beats, it can try to refresh.
         */
        this.refresh = function () {
          if (ws) {
            ws.close();
          }
        };
      }

      /**
       * An event listener to be called when the $window.WebSocket connection's readyState changes to OPEN;
       * this indicates that the connection is ready to send and receive data.
       */
      WebSocket.prototype.onopen = function (event) { };
      /** An event listener to be called when the $window.WebSocket connection's readyState changes to CLOSED. */
      WebSocket.prototype.onclose = function (event) { };
      /** An event listener to be called when a connection begins being attempted. */
      WebSocket.prototype.onconnecting = function (event) { };
      /** An event listener to be called when a message is received from the server. */
      WebSocket.prototype.onmessage = function (event) { };
      /** An event listener to be called when an error occurs. */
      WebSocket.prototype.onerror = function (event) { };

      /**
       * Whether all instances of WebSocket should log debug messages.
       * Setting this to true is the equivalent of setting all instances of WebSocket.debug to true.
       */
      WebSocket.debug_all = false;

      WebSocket.CONNECTING = $window.WebSocket.CONNECTING;
      WebSocket.OPEN = $window.WebSocket.OPEN;
      WebSocket.CLOSING = $window.WebSocket.CLOSING;
      WebSocket.CLOSED = $window.WebSocket.CLOSED;

      return WebSocket;

    }]);
})(angular);